### DISCUSSION TITLE
[UI] Interfacing via outside code and/or an integrated programming language (prefarably lua)

### DISCUSSION THREAD

**Discussion by IoIxD** [olive-editor/olive#2221] [Ideas]
**Summary**

One of the design goals, according to Matt in a recent Patreon post (on his personal account, not the one for Olive) was...

> if there's a niche feature or automated action that helps your workflow, it's significantly easier to add it to an open source program than a closed source program

I think this brings up a good point about the addition of either an integrated programming language, or an interface by which UI and etc. is exposed to a third party program, such as a TCP stream that takes commands as input. There's also the idea of Olive just opening up a DLL/so file and running it while linking internal functions against it.

All are good options and honestly I think implementing both could be a good idea; an integrated language for easy, simple scripting, and an FFI of some sort for more complex libraries. 

If an integrated language IS done, its worth considering which one, btw. Javascript would be easiest to implement since QT has an integrated language already, but Lua is also an option and the one I'd be more fond of, because it's much more portable and embeddable by design, more suited as an integrated language, and most important, more likely to have a complete implementation; Javascript is a humungous language and as far as I've seen the only complete implementations of it are the ones from the three major browser engine vendors; Chrome, Firefox, and Safari. Lua, by contrast, is a very small language thats, again, literally designed to be linked to another program via its hooks system. There's also Python which is a decent option, but as far as I know its a bit harder to integrate into a project and is generally more suited for making standalone applications.

*(It's worth adding, by the way, that the Lua integrated would be original Lua, not the "luau" Roblox uses. I've never used it and can't speak for it personally)*

**Comment by IoIxD** [comment]
After working with lua as an integrated language more i'd like to add that if it were used for Olive, it should instead be luajit, which is violently optimized for performance and adds async functions

**Reply by Simran-B** [reply]
Given the prevalence of Python in the VFX space and Qt's Python support, I think Python is the more likely candidate. OBS on the other hand supports Lua and is also Qt-based. It's not as nice of a language though IMO.

**Reply by IoIxD** [reply]
> Given the prevalence of Python in the VFX space and Qt's Python support, I think Python is the more likely candidate. OBS on the other hand supports Lua and is also Qt-based. It's not as nice of a language though IMO.

It's true that Python is nicer and maybe more complete, but Lua wins in terms of speed which I feel could be important when doing scripting for a video editor.

**Reply by Simran-B** [reply]
I'm not sure if performance is really of relevance. For automating things most likely not. On the other hand, for something like a value-generating node with code that runs on every frame change, it might be.

**Reply by IoIxD** [reply]
Hey, so I actually happened to see something today that you might be interested in? [MoonScript.](https://moonscript.org/)

It's a language that compiles to Lua that adds a lot of modern syntax changes and a real class system (although I imagine it's just an abstraction for Lua tables)
